### PR TITLE
road to v3.0.0

### LINKED ISSUE #24: Gateway creation variables misnamed
From PR #23 review: "If your variable is controlling a behaviour, use imperative style to name it e.g. lockdown_default_seclist"

These variables should be renamed to the suggested name to be in line with this coding convention.
- var.internet_gateway_enabled -> var.create_internet_gateway
- var.nat_gateway_enabled -> var.create_nat_gateway
- var.service_gateway_enabled -> var.create_service_gateway

It is imperative mode, but I also suggest to change the verb from `enable` to `create`.
`enable` and `create` are different verbs from the API point of view, and we should reflect it more precisely, as the current module action is "create in enabled state".

For example, it is totally possible to *create* a nat gateway with a *disabled* state.

Yet, renaming these variables is a breaking change and would need a major version bump.
Plan should be to change these variables on the next major version but start using the suggested naming here for any new variable that controls a behaviour.

### LINKED ISSUE #29: Update examples README
## module can now be used directly from the registry

Step 5 ask to clone the module using git. This module now being live in the terraform registry, we should update the README accordingly

## module composition

the pattern described by "Example reusing terraform-oci-vcn and extending to create other network resources." is known as module composition. We should use this term and point to [the related terraform documentation ](https://www.terraform.io/docs/language/modules/develop/composition.html).

Some required modules for this approach are not existing or not published yet : content will need to evolve using modules as much as possible as they become available.

### LINKED ISSUE #30: Add custom display name for gateways
The module offers currently two choices to name the gateways, but neither is fully customizable:
- prefix+fixed_value
- fixed_value

For example, an internet gateway will be named either internet-gateway or terraform-oci-internet-gateway with prefix set to "terraform-oci" (default value).

This is a feature request to add custom name support for gateways (igw, ngw, sgw, drg) directly from module inputs.
Ideally, these inputs should be optional, with sane defaults.

### LINKED ISSUE #44: Add display_name to drg attachment
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
The DRG Attachment is now visible in the OCI console. Currently, we do not set it so it gets named like "drgattachment20210601034821". Can we please put some formatting so the attachment is neatly named?

### New or Affected Resource(s)

oci_core_drg_attachment.drg 

### Potential Terraform Configuration

```hcl
# Copy-paste any Terraform configurations for how the requested feature may be used. 
```

### LINKED ISSUE #54: rename var.tag to be more specific
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

currently `var.tag` controls the freeform tags creation. We should make it more specific, so we can also support defined tags without confusion.

### New or Affected Resource(s)

all resources managed by the module

### Potential Terraform Configuration

`var.tags`--> `var.freeform_tags`

### LINKED ISSUE #56: Add a validation condition to gateway display name variables
Add a validation condition so the display name in the terraform.tfvars file is not an empty string: https://www.terraform.io/docs/language/values/variables.html#custom-validation-rules

_Originally posted by @hyder in https://github.com/oracle-terraform-modules/terraform-oci-vcn/pull/55#discussion_r698889654_

### LINKED ISSUE #57: simplify the custom route rules variables schema
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Custom route rules feature introduced by fixing #17 in [release v2.2.0](https://github.com/oracle-terraform-modules/terraform-oci-vcn/releases/tag/v2.2.0) use twi new module Input Variables:
- `var.internet_gateway_route_rules`
- `var.nat_gateway_route_rules`

They are both [complex structural types](https://www.terraform.io/docs/language/expressions/type-constraints.html#structural-types).

```
type = list(object({
  destination       = string
  destination_type  = string
  network_entity_id = string
  description       = string
}))
```

Advantages:
- the variable is self-documenting
- we enforce a schema that is expected by OCI API.

Drawbacks:
- it can be complex for new users,
- seems Resource Manager do not handle objects well
- all fields becomes required. [Optional object type attributes is still an experimental feature from 0.14](https://www.terraform.io/docs/language/expressions/type-constraints.html#experimental-optional-object-type-attributes)

The goal is to simplify usage of these new variables, without loosing too much on schema enforcement and documentation.

- solution 1: use type `map(any)`
- solution 2: use type `list(map(any))` or `list(map(string))`
- solution 3: use type `list(object)` with optional attributes

_Solution 3 may be a long term approach, when optional attributes will become a supported feature of Terraform. It gives both self-documented variable and simplified input with optional attributes._

With `map(any)`, the module user would have to provide a key for each rule despite the OCI API do not require it and the accepted schema needs to be discovered by the user from documentation. That's not simplification.

Using `list(map(any))` we we loose some constraint validation in comparison with `list(map(string))`.

I would suggest using `list(map(string))` instead of the current `list(object)` and reevaluate this choice when Optional object type attributes becomes a supported feature.

The change can be transparent for the module users in later releases if we stay with a list type at the primary level.

### Potential Terraform Configuration

```hcl
variable "nat_gateway_route_rules" {
type = list(map(string))
default = [
  {
    destination = "192.168.0.0/16" # Route Rule Destination CIDR
    destination_type = "CIDR_BLOCK"
    network_entity_id = "drg"
    description = "Terraformed - User added Routing Rule"
    },
    {
      destination = "172.16.0.0/16"
      destination_type  = "CIDR_BLOCK"
      network_entity_id = module.vcn.drg_id
      description  = "Terraformed - User added Routing Rule"
    },
  ]
}
```

### PR DESCRIPTION
**Pull request by kral2**
This PR should fix all remaining open points in [milestone v3.0.0](https://github.com/oracle-terraform-modules/terraform-oci-vcn/milestone/3).

# Proposed change

- [x] fix #24
- [x] fix #54
- [x] fix #30
- [x] fix #44
- [x] fix #29 
- [x] fix #56  
- [x] fix #57  

### Manual testing

If no automated testing is run, please ensure that at least the three steps below are passing without any error.

- [x] Running `terraform apply` on each example provided with this module provisions the intended resource(s) without any errors.
- [x] Modifying module's *Input Variables* after initial provisioning behaves as intended, i.e: any updateable properties are ameneded without recreation of the resource(s).
- [x] Running `terraform destroy` on each example provided with this module destroys all the resources created by this module and only the resources created by this module.

## Checklist before submitting your PR

- [x] My code follows [the style guidelines of this project](../tree/main/docs/codingconventions.adoc)
- [x] these changes provision new resources:
  - [x] I have updated the README introduction section (README.adoc)
  - [x] I have updated the README introduction section (README.md)
- [x] these changes adds any new variables:
  - [x] I have updated [docs/terraformoptions.adoc](../tree/main/docs/terraformoptions.adoc) to include each variable
- [x] I have updated the changelog to include an entry for these changes
- [x] I have updated all provided examples, including each README file and all applicable code blocks
- [x] these changes generates no new warnings
- [x] Any dependent changes have been merged and published in upstream modules

Note: *If you are not an Oracle employee, to contribute to an Oracle-sponsored open-source project, you need to sign the [Oracle Contributor Agreement (OCA)](https://oca.opensource.oracle.com/).*

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -102,9 +98,28 @@ variable "vcn_name" {
 variable "drg_display_name" {
```
Comment (hyder) [inline variables.tf:98]: Add a validation condition so the display name in the terraform.tfvars file is not an empty string: https://www.terraform.io/docs/language/values/variables.html#custom-validation-rules

File: variables.tf [hunk 2]
```diff
@@ -102,9 +98,28 @@ variable "vcn_name" {
...
+  default     = "drg"
+}
+
+variable "internet_gateway_display_name" {
+  description = "(Updatable) Name of Internet Gateway. Does not have to be unique."
+  type        = string
+  default     = "internet-gateway"
 }
 
+variable "nat_gateway_display_name" {
```
Comment (hyder) [inline variables.tf:110]: Add a validation condition so the display name in the terraform.tfvars file is not an empty string: https://www.terraform.io/docs/language/values/variables.html#custom-validation-rules

File: variables.tf [hunk 3]
```diff
@@ -102,9 +98,28 @@ variable "vcn_name" {
...
+  description = "(Updatable) Name of NAT Gateway. Does not have to be unique."
+  type        = string
+  default     = "nat-gateway"
+}
+
+variable "service_gateway_display_name" {
```
Comment (hyder) [inline variables.tf:116]: Add a validation condition so the display name in the terraform.tfvars file is not an empty string: https://www.terraform.io/docs/language/values/variables.html#custom-validation-rules

File: variables.tf [hunk 4]
```diff
@@ -102,9 +98,28 @@ variable "vcn_name" {
 variable "drg_display_name" {
   description = "(Updatable) Name of Dynamic Routing Gateway. Does not have to be unique."
   type        = string
-  default     = null
+  default     = "drg"
+}
+
+variable "internet_gateway_display_name" {
```
Comment (hyder) [inline variables.tf:104]: Add a validation condition so the display name in the terraform.tfvars file is not an empty string: https://www.terraform.io/docs/language/values/variables.html#custom-validation-rules

**Review by hyder — CHANGES_REQUESTED** [review]

File: .terraform.lock.hcl [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# This file is maintained automatically by "terraform init".
+# Manual edits may be lost in future updates.
+
+provider "registry.terraform.io/hashicorp/oci" {
+  version     = "4.41.0"
```
[the remark this answers is not in the document]
Comment (kral2) [reply .terraform.lock.hcl:5]: That was a mistake: >= is what we want. I will correct that

I had to bump the version to support some naming options on DRG attachment, but did not found the required version: just took the last one ...

File: vcn_gateways.tf [hunk 6]
```diff
@@ -199,16 +199,19 @@ resource "oci_core_route_table" "nat" {
...
 
   count = var.create_drg == true ? 1 : 0
 }
 
 resource "oci_core_drg_attachment" "drg" {
-  drg_id = oci_core_drg.drg[count.index].id
-  vcn_id = oci_core_vcn.vcn.id
+  drg_id       = oci_core_drg.drg[count.index].id
+  vcn_id       = oci_core_vcn.vcn.id
+  display_name = var.label_prefix == "none" ? "${var.drg_display_name}_to_${oci_core_vcn.vcn.display_name}" : "${var.label_prefix}-${var.drg_display_name}_to_${oci_core_vcn.vcn.display_name}"
```
[the remark this answers is not in the document]
Comment (kral2) [reply vcn_gateways.tf:212]: You mean <name_with_underscore>-to-<name_with_underscore> ?

**Review by hyder — APPROVED** [review]
approved

**Pull request merged**